homeassistant/components/tuya/__init__.py

```diff
@@ -260,18 +163,71 @@ async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
     return unload
 
 
-async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Async setup hass config entry."""
     _LOGGER.debug("tuya.__init__.async_setup_entry-->%s", entry.data)
 
-    hass.data[DOMAIN] = {
-        TUYA_HA_TUYA_MAP: {},
-        TUYA_HA_DEVICES: [],
-        TUYA_SETUP_PLATFORM: set(),
-    }
+    hass.data[DOMAIN] = {TUYA_HA_TUYA_MAP: {}, TUYA_HA_DEVICES: {}}
 
     success = await _init_tuya_sdk(hass, entry)
     if not success:
         return False
 
     return True
+
+
+class DeviceListener(TuyaDeviceListener):
+    """Device Update Listener."""
+
+    def __init__(self, hass: HomeAssistant) -> None:
+        """Init DeviceListener."""
+
+        self.hass = hass
+
+    def update_device(self, device: TuyaDevice) -> None:
+        """Update device status."""
+        if device.id in self.hass.data[DOMAIN][TUYA_HA_DEVICES]:
+            _LOGGER.debug(
+                "_update-->%s;->>%s",
+                self,
+                self.hass.data[DOMAIN][TUYA_HA_DEVICES][device.id].tuya_device.status,
+            )
+            async_dispatcher_send(self.hass, TUYA_HA_SIGNAL_UPDATE_ENTITY)
+
+    def add_device(self, device: TuyaDevice) -> None:
+        """Add device added listener."""
+        device_add = False
+
+        _LOGGER.debug(
+            """add device category->%s; keys->,
+            {hass.data[DOMAIN][TUYA_HA_TUYA_MAP].keys()}""",
+            device.category,
+        )
+
+        if device.category in itertools.chain(
+            *self.hass.data[DOMAIN][TUYA_HA_TUYA_MAP].values()
+        ):
+            ha_tuya_map = self.hass.data[DOMAIN][TUYA_HA_TUYA_MAP]
+
+            self.hass.add_job(async_remove_hass_device, self.hass, device.id)
+
+            for key, tuya_list in ha_tuya_map.items():
+                if device.category in tuya_list:
+                    device_add = True
+                    dispatcher_send(
```

Comment:
Store the device id in a set in `hass.data[DOMAIN][TUYA_HA_DEVICES]` when discovering a new device before dispatching the discovery to add new entities.

Reply:
Why do you want to do this? After I dispatching the discovery to add new entities, it will call _setup_entities, and then Store the device id in a set in hass.data[DOMAIN][TUYA_HA_DEVICES]

Reply:
The dispatch handler will be scheduled so it won't have returned when `dispatcher_send` returns. So we need to track discovered devices here and not in the platform. Otherwise we could potentially send more than one signal for a discovered device.

Reply:
In switch.py, we will receive the message of the newly added device, because different devices need to be judged differently according to the type of device when adding, so we put the logic in the corresponding device file for processing, if it is in the init file Inside, the code will become redundant

![image](https://user-images.githubusercontent.com/22512175/131640973-bc21d65e-1164-4685-87ad-eaecbf7dc205.png)

![image](https://user-images.githubusercontent.com/22512175/131641065-bd3a0266-93dc-4049-9b95-93546b8396c6.png)

Reply:
It's the platform's responsibility to translate the device to the correct entity or entities. It's the device manager's responsibility to keep track of existing or new devices. I don't think we need to change the code you show above.

The only part we need to change is that we should track the discovered device ids in a set stored in hass.data, or somewhere on the device manager.

Reply:
Do you mean to track devices, even if some non-compliant devices are added to the set? Because there is no verification of whether the device can be converted into an entity here, the consequence is that even if some devices cannot be converted into an entity, the id will be stored in the set.

Reply:
Yes. Whether or not a device can be converted to an entity won't change until we release a new version of Home Assistant where that logic is updated. So that detail is not something we need to track. We only need to track existing devices and add new devices to that set when new devices are discovered.

Reply:
ok, thank you for your patience👍